### PR TITLE
Call destructors on goto out of scope

### PR DESCRIPTION
**Pull request by smowton**
CBMC already generates `dead` instructions when an automatic variable leaves scope due to exiting a lexical scope (`{}` block). However a `dead` op is not generated when the block is left by a `goto` instruction. This PR adds support for that, first adding destructors whenever a block is exited, and then restoring simple guarded gotos when possible.

This also adds `dead` instructions in appropriate places if lexical scope inference is done for Java (for which, see https://github.com/diffblue/cbmc/pull/392), as due to the translation from Java bytecode, `goto` statements are a normal way to leave a lexical scope in this case.

### CODE REVIEW

File: src/goto-programs/goto_convert.cpp [hunk 1]
```diff
@@ -183,6 +225,61 @@ void goto_convertt::finish_computed_gotos(goto_programt &goto_program)
...
+          between goto z and z, and if not, simplify into
+          if(!x) goto y;
+
+\*******************************************************************/
+
+void goto_convertt::finish_guarded_gotos(goto_programt &dest)
+{
+  std::map<goto_programt::targett, int> itertoint;
+
+  int i=0;
```
Comment (mgudemann) [inline src/goto-programs/goto_convert.cpp]: reduce scope, change to `size_t`

File: src/goto-programs/goto_convert.cpp [hunk 2]
```diff
@@ -2133,6 +2230,24 @@ void goto_convertt::collect_operands(
...
+
+  Inputs: Goto program 'g'
+
+ Outputs: True if 'g' has one instruction
+
+ Purpose: This is (believed to be) faster than using std::list.size
+
+\*******************************************************************/
+
+static inline bool is_size_one(const goto_programt &g)
```
Comment (mgudemann) [inline src/goto-programs/goto_convert.cpp]: is this optimization necessary? Could this be moved to a class or a util file?

File: src/goto-programs/goto_convert.cpp [hunk 3]
```diff
@@ -2261,6 +2376,14 @@ void goto_convertt::generate_ifthenelse(
   assert(!tmp_w.instructions.empty());
   x->source_location=tmp_w.instructions.back().source_location;
 
+  // See if we can simplify this guarded goto later.
+  // Note this depends on the fact that `instructions` is a std::list
+  // and so goto-program-destructive-append preserves iterator validity.
+  if(is_guarded_goto)
+    guarded_gotos.push_back({tmp_v.instructions.begin(),
+                             tmp_w.instructions.begin(),
+                             guard});
```
Comment (mgudemann) [inline src/goto-programs/goto_convert.cpp]: cpplint

File: src/goto-programs/goto_convert_exceptions.cpp [hunk 4]
```diff
@@ -375,22 +375,32 @@ void goto_convertt::unwind_destructor_stack(
   std::size_t final_stack_size,
   goto_programt &dest)
 {
+  unwind_destructor_stack(source_location,final_stack_size,
+                          dest,targets.destructor_stack);
```
Comment (mgudemann) [inline src/goto-programs/goto_convert_exceptions.cpp]: cpplint

File: src/goto-programs/goto_convert.cpp (hunk 2 above)
Comment (smowton) [inline src/goto-programs/goto_convert.cpp]: No idea, inherited it from the original code -- suggest asking @kroening about this as he was the original author

**Comment by kroening** [comment]
May I suggest adding the obvious test?

**Comment by smowton** [comment]
Addressed Matthias' comments

**Comment by smowton** [comment]
Added test (see commit message for details) and rebased

**Pull request merged**